src/lib/Type.hs

```diff
@@ -186,116 +167,101 @@ checkKindEq k1 k2 | k1 == k2  = return ()
+    FVar v          -> getEffType v
     FDecl decl body -> (eff, bodyTy)
-      where
-        (bodyEff, bodyTy) = getEffType body
-        eff = ignoreExcept $ combineEffects bodyEff $ getFDeclEff decl
+      where (bodyEff, bodyTy) = getEffType body
+            declEff = case decl of
+              LetMono _ rhs -> fst $ getEffType rhs
+              _             -> noEffect
+            eff = ignoreExcept $ combineEffects bodyEff declEff
     FPrimExpr e -> case e of
-      ConExpr con -> pureTy $ getConType $ fmapExpr con id getType getFLamType
-      OpExpr  op  -> getOpType $ fmapExpr op id addType getFLamType
-        where addType x = (fromAtomicFExpr x, getType x)
+      ConExpr con -> pureType $ getConType $ fmapExpr con id getType getPiType
+      OpExpr  op  -> getOpType $ fmapExpr op id getTypeAndAtom getPiType
+      where getTypeAndAtom x = (fromAtomicFExpr x, getType x)
     SrcAnnot e _ -> getEffType e
+    -- The annotation only specifies the type of the result of e,
+    -- not the effects e might produce!
     Annot e _    -> getEffType e
 
   checkEffType expr = case expr of
-    FVar v -> liftM pureTy $ checkVar v
-    FDecl decl body -> do
-      (eff, env) <- checkTypeFDecl decl
-      (eff', ty) <- extendR env $ checkEffType body
-      eff'' <- combineEffects eff eff'
-      return (eff'', ty)
-    FPrimExpr (ConExpr con) -> do
-      eTy <- traverseExpr con return checkPureType checkTypeFlam
-      liftM pureTy $ checkConType eTy
-    FPrimExpr (OpExpr con) -> do
-      eTy <- traverseExpr con return addType checkTypeFlam
-      checkOpType eTy
-      where addType x = liftM (fromAtomicFExpr x,) (checkPureType x)
+    FVar v -> liftM pureType $ checkPureType v
+    FDecl decl cont -> do
+      -- Get the eff of rhs and extend the env with binders
+      (declEff, env) <- case decl of
+          LetMono p body -> do
+            (eff, ty) <- checkEffType body
+            assertEq (getType p) ty "LetMono"
+            return (eff, foldMap lbind p)
+          LetPoly b@(_:>ty) (FTLam tbs qs body) -> do
+            body' <- extendClassEnv qs $ extendR (foldMap tbind tbs) $ checkPureType body
+            let ty' = Forall tbs qs body'
+            assertEq ty ty' "TLam"
+            return (noEffect, lbind b)
+          TyDef tv _ -> return (noEffect, tbind tv)
+      (contEff, ty) <- extendR env $ checkEffType cont
+      eff <- combineEffects declEff contEff
+      return (eff, ty)
+      where
+    FPrimExpr e -> case e of
+      ConExpr con -> do
+        eTy <- traverseExpr con return checkPureType checkPiType
+        liftM pureType $ checkConType eTy
+      OpExpr op -> do
+        eTy <- traverseExpr op return checkTypeAndAtom checkPiType
+        checkOpType eTy
+      where checkTypeAndAtom x = liftM (fromAtomicFExpr x,) (checkPureType x)
     SrcAnnot e pos -> addSrcContext (Just pos) $ checkEffType e
     Annot e ty -> do
-      ty' <- checkPureType e
+      (eff, ty') <- checkEffType e
```

Comment:
By the way this was a bug, but was currently getting masked by the fact that inference drops all annotations!

Reply:
good catch!